core/client/db/src/cache/list_cache.rs

```diff
@@ -512,6 +561,43 @@ impl<Block: BlockT, T: CacheItemT> Fork<Block, T> {
 			best_finalized_block,
 		)
 	}
+
+	/// Truncate fork by deleting all entries that are descendants of given block.
+	pub fn truncate<S: Storage<Block, T>, Tx: StorageTransaction<Block, T>>(
+		&self,
+		storage: &S,
+		tx: &mut Tx,
+		reverting_block: NumberFor<Block>,
+		best_finalized_block: NumberFor<Block>,
+	) -> ClientResult<Option<Fork<Block, T>>> {
+		let mut current = self.head.valid_from.clone();
```

Comment:
I would have see 'self.best_block' here, I understand truncate as a reverse iteration on fork range of blocks deleting in tx all entry until finalization or the intersection of the fork and its parent branch which is the reverting block number).
Maybe I got confuse between 'self.best_block' and 'self.head.valid_from', for me it defines a range of block number for the branch/fork.

Reply:
The `best_block` of the `Fork` struct is an `Option`al field that is used just for optimization - i.e. when new block is inserted into database, we normally are not reading anything from database at all. Instead we are checking if new block is appended to some known fork by checking if `Some(block.parent_hash()) == fork.best_block`. If that's true => then we have found a fork where we going to append a new/same value => no reads from DB are performed. We only access database when some new fork is created - i.e. new block parent isn't the best block of some known fork.

`head.valid_from` points to the block where value (like CT configuration, or authorities, ...) has been updated last time. I.e. it is a normal situation when value has been updated at 100 and the best block (of this fork) is 150. So this loop is exactly what you have said - a reverse iteration. But the iteration elements aren't blocks, but instead nodes of the list-cache.

Reply:
Ok, guess I confuse 'valid_from' with 'prev_valid_from' of 'StorageEntry', I will need to read that again :)

Reply:
The conversion from `StorageEntry` to `Entry` is quite simple. There's a list of all active forks saved in the cache' meta. Every fork is described by the `ComplexBlockId` which points to the `head` of the fork. We then read the value from the head && when converting to `Entry` simply ignore the `prev_valid_from` field of the `StorageEntry`. Instead, value and the block of the `head` is used. wE only use `prev_valid_from` when traversing the list - when truncating or searching for the value that has been actual at the given block, ...